core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java

```diff
@@ -360,6 +365,7 @@ synchronized void updateAllocationIdsFromPrimaryContext(final PrimaryContext pri
                 .allMatch(e -> e.value == SequenceNumbersService.UNASSIGNED_SEQ_NO) : inSyncLocalCheckpoints;
         assert StreamSupport
                 .stream(trackingLocalCheckpoints.spliterator(), false)
+                .filter(e -> !e.key.equals(allocationId)) // during primary relocation a shard can already know its local checkpoint
```

Comment:
I think we should not filter out the allocation id, but keep the method as before and instead assert in the `IndexShard.updateAllocationIdsFromPrimaryContext` method that the local checkpoint that is provided as part of the primary context is equal to the local checkpoint that exists locally.